k2/csrc/determinize.cc

```diff
@@ -0,0 +1,230 @@
+         case of ties on weights, is that a failure here could lead to
+         situations where we didn't advance the base state where we could,
+         leading the number of determinized states to be larger than it could
+         be.
+
+       - Advance the base state if possible.  Each DetState can be represented
+         as a base state and a sequence of symbols from that base state, but
+         if some initial subsequence of that symbol sequence takes us to
+         a unique state then we say the DetState is not normalized.  In that
+         case we need to advance the base state and reduced `seq_len`.
+         If this happens, then the arc sequence which takes us to the new
+         base state will be output to `leftover_arcs`.  When this is done,
+         the 'weight' components of the DetStateElement members also need
+         to be adjusted to remove the weight contribution from those arcs.
+
+     The forward_backward_weight is the weight on the best path through the
+     output determinized FSA that will include this DetState.  It will determine
+     the order of expansion of DetStates and also whether the states are
+     expanded at all (if the pruning beam `beam` is finite).
+     forward_backward_weight is the sum of the forward weight of the base state,
+     plus (the greatest over the DetStateElements, of its `weight` element,
+     plus the backward weight in the input FSA of the state that corresponds
+     to it).
+
+
+     worked outobtained from
+
+   */
+  void Normalize(std::vector<int32_t> *leftover_arcs);
+};
+
+
+void DetState::Normalize(std::vector<int32_t> *input_arcs) {
+
+}
+
+
+class DetStateMap {
+ public:
+  /*
+    Outputs the output state-id corresponding to a specific DetState structure.
+    This does not store any pointers to the DetState or its contents, so
+    you can delete the DetState without affecting this object's ability to map
+    an equivalent DetState to the same state-id.
+
+       @param [in] a  The DetState that we're looking up
+       @param [out] state_id  The state-index in the output FSA
+                      corresponding to this DetState (will
+                      be freshly allocated if an equivalent of
+                      this DetState did not already exist.
+        @return  Returns true if this was a NEWLY CREATED state,
+              false otherwise.
+   */
+  bool GetOutputState(const DetState &a, int32_t *state_id) {
+    std::pair<uint64_t, uint64_t> compact;
+    DetStateToCompact(a, &compact);
+    auto p = map_.insert({compact, cur_output_state));
+    bool inserted = p.second;
+    if (inserted) {
+      *state_id = cur_output_state_++;
+      return true;
+    } else {
+      *state_id = p.first->second;
+      return false;
+    }
+  }
+
+  size_t size() const { return cur_output_state_; }
+
+ private:
+
+  int32_t cur_output_state_ { 0 };
+  std::unordered_map<std::pair<uint64_t, uint64_t>, int32_t, DetStateVectorHasher> map_;
```

Comment:
`DetStateHasher`?

Reply:
Yeah, might do that.  This code needs to be finished.